File: discovery/web/ruscertbyaddress.py
#!/usr/bin/env python
#
#   hostmap
#
#   Author:
#    Alessandro `jekil` Tanasi <alessandro@tanasi.it>
#
#   License:
#    This program is private software; you can't redistribute it and/or modify
#    it. All copies, included printed copies, are unauthorized.
#    
#    If you need a copy of this software you must ask for it writing an
#    email to Alessandro `jekil` Tanasi <alessandro@tanasi.it>



import re
from twisted.web import client
from lib.core.outputDeflector import log



class ruscertbyaddress:
    """ 
    Check against ruscert
    @author: Alessandro Tanasi
    @license: Private software
    @contact: alessandro@tanasi.it
    """



    def require(self):
        """
        Sets plugin requirements
        """
        
        # Possible values are:
        # ip
        # domain
        # nameserver
        # hostname
        return "ip"



    def run(self, hd, ip):
        """
        Query RUS CERT-DB to get a list of domains
        """
        
        self.job = "%s-%s" % (__name__, ip)
        hd.job(self.job, "starting")
        
        # Compose url
        url = "http://www.bfk.de/bfk_dnslogger.html?query=%s" % ip
        
        # Search
        query = client.getPage(url)
        query.addCallback(self.__callSuccess, hd)
        query.addErrback(self.__callFailure, hd)
        
        hd.job(self.job, "waiting")
    


    def __callFailure(self, failure, hd):
        """
        If a live search run fails
        """
        
        failure.printTraceback()
        
        hd.job(self.job, "failure")


    def __callSuccess(self, success, hd):
        """
        If a live search run success
        """
        
        # Regexp to catch fqdn
        regexp = "\">([\d\w\.-_]+)</a></tt></td><td><tt>"
        # Cast object, paranoid mode
        page = str(success)
        # Grep
        results = re.findall(regexp, page)
       # Add new found hosts
        for host in set(results):
            hd.notifyHost(host)
            log.debug("Plugin %s added result: %s" % (__name__, host))
            
        hd.job(self.job, "done")

File: discovery/web/ruscertbyaddress.rb
require 'open-uri'
require 'set'

#
# Check against RUS CERT-DB.
#
PlugMan.define :ruscertbyaddress do
  author "Alessandro Tanasi"
  version "0.2.1"
  extends({ :main => [:ip] })
  requires []
  extension_points []
  params({ :description => "Check against RUS CERT-DB" })

  def run(ip, opts = {})
    @hosts = Set.new

    begin
      page = open("http://www.bfk.de/bfk_dnslogger.html?query=#{ip}").read
	  #AAS modificata regular expression
      #page.scan(/\">([\d\w\.-_]+)<\/a><\/tt><\/td><td><tt>/).each do |url|
      page.force_encoding("ISO-8859-1").encode("utf-8", replace: nil).scan(/<tr id="color[0-9]+"><td><tt><a href="http:\/\/www.bfk.de\/bfk_dnslogger.html\?query=.*?#result" rel="nofollow">(.*?)<\/a><\/tt><\/td>/).each do |url|
        @hosts << { :hostname => url[0].to_s } 
      end
    rescue Exception => e
	  puts ":ruscertbyaddress error: #{e.inspect}"
      return @hosts
    end

    return @hosts
  end

  def timeout
    return @hosts
  end
end
